057.- Curso C++ BÃ¡sico. Halts. Salidas anticipadas de un programa
===

El ultimo grupo de declaraciones de control de flujo que veremos son los halts, paradas o detenciones. aunque seria salida la palabra que mejor define lo que hacen porque son declaraciones de control de flujo que finalizan el programa.

 En C++ las paradas se implementan como funciones en lugar de como palabras clave por lo que nuestras declaraciones halt seran llamadas a funciones.

 Cuando un programa finaliza normalmente, es decir cuando la funcion main devuelve un return ya sea porque ha llegado al final de la funcion o por una llamada a return en el codigo, cuando esto ocurre suceden varias cosas:

 1) Como se termina una funcion, todas las variables y parametros locales se destruyen tal como ocurre con cualquier funcion. Y a continuacion se llama a una funcion especial `std::exit(status_code)`.

 std::exit es una funcion que hace que el programa finalice normalmente, es decir ha terminado de forma esperada independientemente de si ha producido algun error o no, por ejemplo: supongamos que estamos escribiendo un archivo en el que se espera que el usuario introduzca un nombre de archivo para procesar, si el usuario escribe un nombre de archivo no valido el programa probablemente devolveria un valor distinto de cero como status_code para indicar el estado del fallo, pero aun asi se seguiria considerando como una terminacion normal.

 std::exit lleva a cabo una serie de funciones de limpieza:
 1) Los objetos con una duracion de almacenamiento estatica se destruyen.

 2) Se realiza una limpieza de archivos innecesarios

 3) Se devuelve el control al sistema operativo con el argumento que se pasa a std::exit() que es utilizado como status code.


Aunque std::exit se llama implicitamente cuando la funcion main finaliza, tambien podemos llamarla explicitamente para detener el programa antes de termine normalmente. Cuando se llame a std::exit de esta manera tenemos que anadir el include a `cstdlib` 

ejemplo
```cpp
#include <cstdlib>
#include <iostream>

void limpiar(){
   // codigo aqui que lleva a cabo cualquier tipo de limpieza necesaria.
   std::cout << "Se realiza la limpieza aqui\n";
}

int main(){

   std::cout << 1 << '\n';
   limpiar();
   std::exit(0); // termina y devuelve el status_code 0 al sistema operativo

   // la declaracion siguiente nunca llega a ser ejecutada
   std::cout << 2 << '\n';
   return 0;
}
```

Sobre el uso de std::exit es importante destacar que:

1)  cuando se hace una llamada a std::exit explicitamente esto no elimina ni las variables locales ni la pila para otras funciones, ya que se esta interfiriendo a la llamada implicita que ocurre cuando se llama al return la limpieza debe asumirla tambien uno mismo. 

2) Las declaraciones posteriores a la llamada std::exit  nunca se ejecutaran porque el programa se cierra en ese momento.

3) Las llamadas a std::exit pueden hacerse desde cualquier funcion o cualquier parte del programa, no tiene porque ser desde  la funcion main.


Es importante limpiar la informacion extra que ha sido creada por el programa, pero tener que anadir la funcion limpiar cada vez que se use exit puede agregar peso y complejidad al programa. PAra evitar este tipo de problemas C++ nos provee otra funcion funcion: `std:atexit()` que permite especificar una funcion que se llamara automaticamente y luego finaliza el programa como lo haria exit. Importante destacar que lo que std::atexit recibe como argumento es una funcion en si misma y no una llamada a una funcion por eso el nombre de la funcion no lleva parentesis asociados, vemos un ejemplo


```cpp
#include <cstdlib>
#include <iostream>

void limpiar(){
   // codigo aqui que lleva a cabo cualquier tipo de limpieza necesaria.
   std::cout << "Se realiza la limpieza aqui\n";
}

int main(){

   std::atexit(limpiar);   // registramos limpiar para que sea llamada automaticamente cuando sea llamada std::exit()
                           // observa que usamos `limpiar` en lugar de `limpiar()` porque no estamos haciendo una llamada
                           // a la funcion limpiar en este momento.

   std::cout << 1 << '\n';
   std::exit(0); // termina y devuelve el status code 0 al sistema operativo

   // la declaracion siguiente nunca se ejecuta.
   std::cout << 2 << '\n';

   return 0;
}
```

Cuando se traba un programa con subprocesos multiples la llamada explicita std::exit o std::atexit puede bloqueear el programa. Dijimos que cuando llamamos a exit explicitamente no se eliminan ni las variables locales ni la pila de llamada de otras funciones pero si se eliminan las variables estaticas, que puede que sigan siendo necesarias para otros subprocesos y puede provocar bloqueos. Por este motivo C++ ha introducido otras 2 funciones llamadas std::quick_exit y std::at_quick_exit que terminan el programa normalmente pero no limpian los objetos estaticos, y podemos decidir si realizar o no otros tipos de limpieza.

C++ tiene otros tipos de funciones relacionadas con salidas del programa. La llamada a `std::abort()` hace que el programa finalice pero de forma anormal, Significa que el programa tuvo algun tipo de error de ejecucion inusual y el programa no pudo continuar ejecutandose. Por ejemplo realizar una division entre cero. std::abort no hace ninguna limpieza. A continuacion vemos un ejemplo de llamada a std::abort



```cpp
#include <cstdlib>
#include <iostream>


int main(){
   std::cout << 1 << '\n';
   std::abort();

   // la siguiente declaracion no se ejecutara nunca
   std::cot << 2 << '\n';
   return 0;
}

```

La funcion std::terminate se usa en la mayoria de los casos con excepciones.

